kvserver/rangefeed: add cluster setting RangefeedUseBufferedSender

wenyihu6 opened the pull request:
This patch adds a cluster setting RangefeedUseBufferedSender. If enabled,
rangefeed will use unbuffered registrations - registrations will not buffer
events at rangfeed level. Instead, registrations will use a BufferedSender
to buffer events at a node level. Note that BufferedSender is left
unimplemented in this patch, and the setting is disabled everywhere. Using it
will cause a panic. We will implement BufferedSender in a follow-up patch.

This patch also introduces other components to make future commits cleaner.

The setting is left undocumented in release note since the feature is not
complete yet.

Part of: https://github.com/cockroachdb/cockroach/issues/129813
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/129228)
<!-- Reviewable:end -->

nvb reviewed (commented):
> Reviewed 7 of 7 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129228)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`pkg/kv/kvserver/rangefeed/buffered_stream_sender.go` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/129228#-O4kTc9e3DY9MjBNvhUu:-O4kTc9e3DY9MjBNvhUv:b-z77x4f) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a6f65a80d212c8959e2c0e0b10ebf4b0dbd4323/pkg/kv/kvserver/rangefeed/buffered_stream_sender.go#L18)):*
> ```Go
> )
> 
> // BufferedStreamSender is a wrapper around ServerStreamSender which buffers
> ```

I'm finding the distinction between "Stream", "Sender", and "Sink" to be not very clear, which is making this object hierarchy difficult to keep straight. Do they all have the same meaning? Do some operate on `MuxRangeFeedEvent` and some operate on `RangeFeedEvent`? Is there something we can do to make the relationship between these concepts more clear?
___
*[`pkg/kv/kvserver/rangefeed/buffered_stream_sender.go` line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/129228#-O4kSERz1sCv3HkjaK-p:-O4kSERz1sCv3HkjaK-q:bsup9o2) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a6f65a80d212c8959e2c0e0b10ebf4b0dbd4323/pkg/kv/kvserver/rangefeed/buffered_stream_sender.go#L30)):*
> ```Go
> // stopped. If it does not return an error, the event is guaranteed to be sent
> // unless the BufferedStreamSender is stopped before sending it (such as due to
> // broken grpc stream). It does not check for any stream-specific information.
> ```

"due to a broken"
___
*[`pkg/kv/kvserver/rangefeed/buffered_stream_sender.go` line 34 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/129228#-O4kSLH74NQKdEuk4K0X:-O4kSLH74NQKdEuk4K0Y:bo5uol1) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a6f65a80d212c8959e2c0e0b10ebf4b0dbd4323/pkg/kv/kvserver/rangefeed/buffered_stream_sender.go#L34)):*
nit: It's strange to see a comment talk about "now" and "before". A reader of the code won't always have the context on how things used to be, so better to just make statements about the current state of the code. Maybe something like:
> // BufferedStreamSender is responsible for using and releasing budget from a SharedBudgetAllocation.
___
*[`pkg/kv/kvserver/rangefeed/stream.go` line 33 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/129228#-O4kStdm9uOdQAtAY3VF:-O4kStdm9uOdQAtAY3VG:b-v4n3ha) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a6f65a80d212c8959e2c0e0b10ebf4b0dbd4323/pkg/kv/kvserver/rangefeed/stream.go#L33)):*
> ```Go
> // BufferedStream is a Stream that can buffer events before sending them to the
> // underlying Stream. Note that the caller may still choose to bypass the buffer
> // and send to the underlying Stream directly by calling Send directly.
> ```

Can we mention that doing so can result in message reordering?
___
*[`pkg/kv/kvserver/rangefeed/stream.go` line 89 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/129228#-O4kU_poCnmGH1PFMXju:-O4kU_poCnmGH1PFMXjv:b-wucvox) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a6f65a80d212c8959e2c0e0b10ebf4b0dbd4323/pkg/kv/kvserver/rangefeed/stream.go#L89)):*
> ```Go
> }
> 
> // BufferedPerRangeEventSink is an implementation of BufferedStream which
> ```

Why add this wrapper type, instead of just adding a `SendBuffered` method directly on `PerRangeEventSink`?


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129228)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)
  
___
*[`pkg/kv/kvserver/rangefeed/buffered_stream_sender.go` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/129228#-O4kTc9e3DY9MjBNvhUu:-O4kd79Z8uD17JqAAHXn:bo42i1z) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a6f65a80d212c8959e2c0e0b10ebf4b0dbd4323/pkg/kv/kvserver/rangefeed/buffered_stream_sender.go#L18)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm finding the distinction between "Stream", "Sender", and "Sink" to be not very clear, which is making this object hierarchy difficult to keep straight. Do they all have the same meaning? Do some operate on `MuxRangeFeedEvent` and some operate on `RangeFeedEvent`? Is there something we can do to make the relationship between these concepts more clear?
</blockquote></details>

I share the sentiment when I was writing these comments. It gets more confusing with the concept of muxstream + more interfaces I'm introducing. I think in general, we have two types of streams - a node level stream (server grpc stream) and a stream from the rangefeed package `type Stream interface`. I usually refer Sender as a server level grpc stream "Sender". 

<img src="https://files.reviewableusercontent.io/images/reviewable/56973754/Kzrs3nKkwVCr7maCM2Tw3ETd/image%20%281%29.png" width="400">

- We use interface for ServerStreamSender to abstract lockedMuxStream and BufferedStreamSender for StreamMuxer. 

- We use interface for rangefeed.Stream and rangefeed.BufferedStream to abstract different sink implementations. We need to distinguish them because we want buffered registrations with rangefeed.Stream and unbuffered registrations with rangefeed.BufferedStream. 

What do think of these new names - does it look better? 
- node level concept 
lockedMuxStream -> NodeMuxStream 
ServerStreamSender -> NodeStream
BufferedStreamSender -> BufferedNodeStream

- rangefeed level concept 
rangefeed.Stream -> rangefeed.PerRangeStream
rangefeed.BufferedStream -> rangefeed.PerRangeBufferedStream 
rangefeed.PerRangeEventSink (leave as it is)
rangefeed.BufferedPerRangeEventSink (leave as it is)
___
*[`pkg/kv/kvserver/rangefeed/stream.go` line 89 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/129228#-O4kU_poCnmGH1PFMXju:-O4khRWd7fde8L5jWHMW:bgrvoiz) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a6f65a80d212c8959e2c0e0b10ebf4b0dbd4323/pkg/kv/kvserver/rangefeed/stream.go#L89)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why add this wrapper type, instead of just adding a `SendBuffered` method directly on `PerRangeEventSink`?
</blockquote></details>

I did it so that we can pass in a different structs to Stores.Rangefeed. Inside p.Register, we can now assert on the type of the Stream being passed in (BufferedStream v.s. regular Stream). Do you see a cleaner way to handle this? It does feel like I'm introducing complexity with abstractions.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129228)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`pkg/kv/kvserver/rangefeed/buffered_stream_sender.go` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/129228#-O4kTc9e3DY9MjBNvhUu:-O4l47hQCFX0kV4M8ZCO:b-48jg15) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a6f65a80d212c8959e2c0e0b10ebf4b0dbd4323/pkg/kv/kvserver/rangefeed/buffered_stream_sender.go#L18)):*
> we have two types of streams - a node level stream (server grpc stream) and a stream from the rangefeed package

> We need to distinguish them because we want buffered registrations with rangefeed.Stream and unbuffered registrations with rangefeed.BufferedStream.

> rangefeed.Stream -> rangefeed.PerRangeStream

In all three of these cases, we're letting the internal details of an interface's implementation dictate its external interface. That's generally not how polymorphism is used. Instead, an interface typically defines the minimum contract that one component requires of another component. Any component which is able to satisfy that interface can then be used as an implementation of it. By defining the minimum contract, we facilitate flexibility and loose coupling. It is then up to the logic which connects these components to compose them in a way that provides its desired properties.

As a concrete example, why would a rangefeed registration care that it holds a `PerRangeStream` vs. a `Stream`? Why would it care whether the stream has an internal buffer or not? Why would it care that it's publishing at the node level or at a range level?

I'm sure I'm missing something, but at a first approximation, it feels like we should have exactly two interfaces: `rangefeed.Stream` and `rangefeed.MuxStream`. `Stream` has a `Send(RangeFeedEvent)` method and `MuxStream` has a `Send(MuxRangeFeedEvent)`. We probably can't actually stick to two because there are additional requirements which users of the interfaces need, like `BufferedSend`. But I think the rule of thumb that "we shouldn't need a separate interface for implementation details that are immaterial to users of the interface" still holds.

----

Separately, I would suggest that we avoid mixing/combining "stream", "sink", and "sender" if they all have the same meaning.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/kv/kvserver/rangefeed/buffered_stream_sender.go` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/129228#-O4kTc9e3DY9MjBNvhUu:-O4lMAS74xVd4dSXTX8X:bsd761c) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a6f65a80d212c8959e2c0e0b10ebf4b0dbd4323/pkg/kv/kvserver/rangefeed/buffered_stream_sender.go#L18)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> we have two types of streams - a node level stream (server grpc stream) and a stream from the rangefeed package

> We need to distinguish them because we want buffered registrations with rangefeed.Stream and unbuffered registrations with rangefeed.BufferedStream.

> rangefeed.Stream -> rangefeed.PerRangeStream

In all three of these cases, we're letting the internal details of an interface's implementation dictate its external interface. That's generally not how polymorphism is used. Instead, an interface typically defines the minimum contract that one component requires of another component. Any component which is able to satisfy that interface can then be used as an implementation of it. By defining the minimum contract, we facilitate flexibility and loose coupling. It is then up to the logic which connects these components to compose them in a way that provides its desired properties.

As a concrete example, why would a rangefeed registration care that it holds a `PerRangeStream` vs. a `Stream`? Why would it care whether the stream has an internal buffer or not? Why would it care that it's publishing at the node level or at a range level?

I'm sure I'm missing something, but at a first approximation, it feels like we should have exactly two interfaces: `rangefeed.Stream` and `rangefeed.MuxStream`. `Stream` has a `Send(RangeFeedEvent)` method and `MuxStream` has a `Send(MuxRangeFeedEvent)`. We probably can't actually stick to two because there are additional requirements which users of the interfaces need, like `BufferedSend`. But I think the rule of thumb that "we shouldn't need a separate interface for implementation details that are immaterial to users of the interface" still holds.

----

Separately, I would suggest that we avoid mixing/combining "stream", "sink", and "sender" if they all have the same meaning.
</blockquote></details>

Hm, I see what you mean. I think it gets complicated because the stream we are passing into rangefeed decides if unbuffered v.s. buffered registrations should be used. We can also just introduce a boolean variable and simply check the boolean. 

I originally thought about something like this below, but thought it was also not as clear as I want it to be: 

- At node level, we only have rangefeed.MuxStream. It is implemented by struct lockedMuxStream (which is unbuffered) and struct BufferedMuxStream (which is buffered and simply wraps lockedMuxStream).

- StreamMuxer will store rangefeed.MuxStream which may be lockedMuxStream or BufferedMuxStream.

- We pass in rangefeed.PerRangeEventStream (implements the rangefeed.Stream interface) to rangefeed. 

- rangefeed.Stream now needs to have another method declaration - ShouldUseBufferedRegistration() to tell p.Register() which type of registrations to use. 

- We also need to change the Send method on the rangefeed.Stream to include another boolean parameter indicating whether the event should be buffered. We will pass in this boolean all the way to StreamMuxer. StreamMuxer will do a switch case and assert its rangefeed.MuxStream is BufferedMuxStream if they want to Send with buffered = true. 

- We also need to pass in `*SharedBudgetAllocation` to the Send method on rangefeed.Stream since buffered queue is responsible for using and releasing the alloc. Buffered registration can just pass in nil for it. 

- We also need to add a RegisterRangefeedCleanUp method in rangefeed.Stream (will be introduced in future commit https://github.com/cockroachdb/cockroach/pull/129231/commits/fd2c751b391c06774972568ca13f000687c8f807). This is only used for unbuffered registration and is needed to handle rangefeed level cleanup since they don't have a dedicated goroutine to do unregister from processor. But adding this will enforce all test structs to implement this additional noop method. We should probably move `RangeFeedEventSink` away from kvpb package if it takes in alloc. 

Do you feel this one is better? lmk if you prefer to facilitate the discussion on slack instead. 


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129228)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`pkg/kv/kvserver/rangefeed/buffered_stream_sender.go` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/129228#-O4kTc9e3DY9MjBNvhUu:-O4mdFGQ1PO3U32wgDbt:b-ksampx) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a6f65a80d212c8959e2c0e0b10ebf4b0dbd4323/pkg/kv/kvserver/rangefeed/buffered_stream_sender.go#L18)):*
> As a concrete example, why would a rangefeed registration care that it holds a `PerRangeStream` vs. a `Stream`? Why would it care whether the stream has an internal buffer or not? Why would it care that it's publishing at the node level or at a range level?

Rangefeed registration shouldn't care if it holds range level `PerRangeStream` vs. node level`MuxStream`. p.Register should only take in a `PerRangeStream`. However, it will care whether the stream has an internal buffer. 

1. Using unbuffered registration is enabled by the StreamMuxer having a buffered sender. Alternatively, we can also check the cluster setting `RangefeedUseBufferedSender` again at the rangefeed level. Ideally, I would want the cluster setting to be checked at only one place tho. 
2.  Unbuffered registration has the flexibility to send data either buffered or unbuffered to the stream. I’m allowing catch-up scans to bypass the buffer and send data directly as unbuffered in the future commit. I think technically, we can just make it send to the buffered queue as well. I did it this way since I was worried catch up scans sending too many events and overflowed too quickly. Currently, they are not subject to memory accounting on master branch, and I'm not sure how large they can grow if I start sending to queue. 
3. SendBuffered have different function signature than SendUnbuffered as it needs to pass `*SharedBudgetAllocation`. 
4. Unbuffered registration must invoke RegisterRangefeedCleanUp with StreamMuxer since it depends on the StreamMuxer for additional cleanup tasks.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129228)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`pkg/kv/kvserver/rangefeed/buffered_stream_sender.go` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/129228#-O4kTc9e3DY9MjBNvhUu:-O4pWJRW6hXa-l14NSJW:bkvjqqf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a6f65a80d212c8959e2c0e0b10ebf4b0dbd4323/pkg/kv/kvserver/rangefeed/buffered_stream_sender.go#L18)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

> As a concrete example, why would a rangefeed registration care that it holds a `PerRangeStream` vs. a `Stream`? Why would it care whether the stream has an internal buffer or not? Why would it care that it's publishing at the node level or at a range level?

Rangefeed registration shouldn't care if it holds range level `PerRangeStream` vs. node level`MuxStream`. p.Register should only take in a `PerRangeStream`. However, it will care whether the stream has an internal buffer. 

1. Using unbuffered registration is enabled by the StreamMuxer having a buffered sender. Alternatively, we can also check the cluster setting `RangefeedUseBufferedSender` again at the rangefeed level. Ideally, I would want the cluster setting to be checked at only one place tho. 
2.  Unbuffered registration has the flexibility to send data either buffered or unbuffered to the stream. I’m allowing catch-up scans to bypass the buffer and send data directly as unbuffered in the future commit. I think technically, we can just make it send to the buffered queue as well. I did it this way since I was worried catch up scans sending too many events and overflowed too quickly. Currently, they are not subject to memory accounting on master branch, and I'm not sure how large they can grow if I start sending to queue. 
3. SendBuffered have different function signature than SendUnbuffered as it needs to pass `*SharedBudgetAllocation`. 
4. Unbuffered registration must invoke RegisterRangefeedCleanUp with StreamMuxer since it depends on the StreamMuxer for additional cleanup tasks.
</blockquote></details>

Thanks for explaining all of this! It does sound like there's a pretty compelling reason for a `SendBuffered` method to exist, given the different method signature. So I think what you have in this PR with extension interfaces (e.g. `BufferedStream`, `BufferedMuxStream`) and extension implementations (e.g. `BufferedPerRangeEventSink`) makes sense. That also gives us the flexibility to extend these interfaces with the `CleanUp` method later on.

If we're going to go with that approach, shouldn't we go all the way and avoid any "runtime" type casts? The one that stands out is `StreamMuxer.SendBuffered`.

I'll reach out on Slack so we can continue to discussion.


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
kvserver/rangefeed: add cluster setting RangefeedUseBufferedSender

This patch adds a cluster setting RangefeedUseBufferedSender. If enabled,
rangefeed will use unbuffered registrations - registrations will not buffer
events at rangfeed level. Instead, registrations will use a BufferedSender
to buffer events at a node level. Note that BufferedSender is left
unimplemented in this patch, and the setting is disabled everywhere. Using it
will cause a panic. We will implement BufferedSender in a follow-up patch.

This patch also introduces other components to make future commits cleaner.

The setting is left undocumented in release note since the feature is not
complete yet.

Part of: https://github.com/cockroachdb/cockroach/issues/129813
Release note: none

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 10 of 10 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129228#-:-O5U1s-s9mM38w5ovtmK:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`pkg/server/node.go` line 1950 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/129228#-O5U1Hj97FREtsxqvM-q:-O5U1Hj97FREtsxqvM-r:b-wd48pi) ([raw file](https://github.com/cockroachdb/cockroach/blob/4714831d75ce364fda9ebd6711b9ab8269180c9f/pkg/server/node.go#L1950)):*
> ```Go
> 	// SendBufferedError disconnects the stream with the ev.StreamID and sends
> 	// error back to client. This call is un-blocking, and additional clean-up
> 	// takes in place async. Caller cannot expect immediate disconnection.
> ```

s/takes in place/takes place/
___
*[`pkg/server/node.go` line 1980 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/129228#-O5U1RTV9A-4JqV9EuwE:-O5U1RTV9A-4JqV9EuwF:b-91r5od) ([raw file](https://github.com/cockroachdb/cockroach/blob/4714831d75ce364fda9ebd6711b9ab8269180c9f/pkg/server/node.go#L1980)):*
> ```Go
> 	} else {
> 		sm = rangefeed.NewUnbufferedSender(lockedMuxStream, n.metrics)
> 		if err := sm.Start(ctx, n.stopper); err != nil {
> ```

Should we move this out of the `else` branch so that we only call `Start` and `Stop` in one place, regardless of the underlying type of `sm`?


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129228)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten)

___
*[`pkg/server/node.go` line 1950 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/129228#-O5U1Hj97FREtsxqvM-q:-O5U9ixk8LbcgsYScXFi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4714831d75ce364fda9ebd6711b9ab8269180c9f/pkg/server/node.go#L1950)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

s/takes in place/takes place/
</blockquote></details>

Done.
___
*[`pkg/server/node.go` line 1980 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/129228#-O5U1RTV9A-4JqV9EuwE:-O5U9jkB7tc5lk7PFr-8:bb77uf3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4714831d75ce364fda9ebd6711b9ab8269180c9f/pkg/server/node.go#L1980)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we move this out of the `else` branch so that we only call `Start` and `Stop` in one place, regardless of the underlying type of `sm`?
</blockquote></details>

Yes!


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129228#-:-O5UDfPnCSoa2LQ3Thes:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @wenyihu6)



<!-- Sent from Reviewable.io -->

wenyihu6 commented:
TFTR! 

bors r=nvanbenschoten

wenyihu6 commented:
bors retry
